proposals/p6333.md

```diff
@@ -0,0 +1,586 @@
+    -   [Implicit carbon_library dependency on Core versus prelude](#implicit-carbon_library-dependency-on-core-versus-prelude)
+    -   [Create an explicit mapping from packaging directives to files](#create-an-explicit-mapping-from-packaging-directives-to-files)
+
+<!-- tocstop -->
+
+## Abstract
+
+-   Change the look-and-feel of the `carbon` compilation command set to use
+    `compile`, `link`, and `build`.
+-   Rewrite `carbon_library` and `carbon_binary` based on the `compile` and
+    `link` commands.
+-   Build library-to-file discovery for `Core`, but support it in a general
+    manner.
+
+## Problem
+
+Right now, `carbon compile` produces an object file per input file. This has a
+few issues, for example when it comes to compiling prelude files: it's not clear
+how to get a prelude `.o` file.
+
+Additionally, the compilation setup doesn't work well with the prelude because
+symbols don't get emitted for prelude files. Right now, we generally don't
+notice this because the prelude has been implemented in a way that typically
+doesn't require separate object files, but it will likely become an issue.
+
+Essentially, we have a decent setup for testing, but not one that's easy to use
+in real-world situations.
+
+## Background
+
+In C++, `clang++ main.cpp` is a way to produce `program`. This is trying to
+reach a similar goal to make it easy to build and test small programs.
+
+Key commands related to this proposal are `carbon compile`, `carbon clang`, and
+`carbon link`. The end result will likely compose multiple command elements in
+order to build the output.
+
+## Proposal
+
+Restructure compilation into:
+
+-   `carbon compile`: Take a single input to build, and produce a single output
+    `.o`.
+-   `carbon build`: Take multiple inputs in order to produce a linked binary.
+    -   Overlaps with `carbon compile` and `carbon link`.
+
+Provide Bazel rules based on this:
+
+-   `carbon_library` similar to `cc_library`.
+-   `carbon_binary` similar to `cc_binary`.
+-   In the `core` directory, `carbon_library` rules that assist with builds.
+    -   The prelude would be a default dependency of all `carbon_library` rules.
+
+These are intended to accept flexible inputs:
+
+-   Support passing in standard C++ file extensions to any of these for
+    compilation.
+-   For `carbon build` in particular, it should not be necessary to pass in
+    `Core` files that are required.
+    -   We will require a correlation between library names inside `Core` and
+        directory structure. For example, `prelude/types`
+        [maps to](#mapping-packaging-directives-to-filenames)
+        `core/prelude/types.carbon`.
+    -   The same strict correlation will be supported for other packages.
+
+At the end, it should be possible to:
+
+-   Run `carbon build main.carbon` (or even just `carbon build`) with
+    non-prelude `Core` imports, and get an executable `program`.
```

Comment:
Does the use of `code font` here for `program` imply that our default output name is `program` (rather than the traditional `a.out`) -- that is, an executable *named* `program`, not just an executable program? I would be fine with taking the base name of one of the input files as the default output name (eg, the file that defines `Run`), or with not having a default, but I don't think we should have a default output file name that doesn't depend on the input name.

Reply:
`program` was used per request from chandlerc:

https://docs.google.com/document/d/19UvmU0znIFDj32hMj7TvE_WkZ_zEHygQHfOiFELKiMU/edit?disco=AAABfcNoIWQ

I think saying ``get an executable `main` `` would also be confusing because "main" is also a term... I've dropped the backticks, do you think that's okay?